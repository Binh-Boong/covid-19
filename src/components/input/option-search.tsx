import { ICountrySelect } from "../../types/apis/country-select.module";

export const OptionSearch = ({
  countries,
  onClick,
}: {
  countries: ICountrySelect[],
  onClick: (param: ICountrySelect) => void,
}): JSX.Element => {
  return (
    <div className="absolute -mt-1 overflow-y-auto bg-white border border-gray-200 border-solid rounded shadow left-1 right-1 max-h-96">
      {countries.length === 0 ? (
        <div className="w-full p-4 text-xl text-center text-gray-500">Empty</div>
      ) : countries.map((item, i) => (
        <div
          className="w-full px-3 py-2 text-lg text-left text-gray-600 cursor-pointer hover:bg-gray-200" key={i.toString()}
          onClick={() => {
            onClick(item);
          }}
          onKeyPress={() => {
            onClick(item);
          }}
          role="button"
          tabIndex={0}
        >
          {item.Country}
        </div>
      ))}
    </div>
  );
}