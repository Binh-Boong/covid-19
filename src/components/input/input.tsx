import React, { useState, useEffect } from "react";

import { IInput, IInputOption } from "../../types/components/input.module";
import { ICountrySelect } from "../../types/apis/country-select.module";
import { OptionSearch } from "./option-search";

export const Input  = ({
  value,
  option,
  onChange,
}: IInput): JSX.Element => {
  const [isVisible, setIsVisible] = useState<boolean>(false);
  const [search, setSearch] = useState<string>(value);
  const [options, setOptions] = useState<ICountrySelect[] | undefined>(option.searchOptions);

  const eventOnChange = (param: string): void => {
    setSearch(param);

    if (option.searchOptions) {
      const optionSearch = option.searchOptions.filter(element => element.Country.trim().toLowerCase().includes(param.trim().toLowerCase()));
      setOptions(optionSearch);
    } else {
      onChange(param);
    }
  }

  return (
    <div className="relative w-full">
      <input
        className={option.className ? option.className : "w-full px-3 py-1 text-gray-700 border border-gray-300 border-solid rounded outline-none focus:border-purple-700 focus:shadow-lg"} 
        value={search} 
        placeholder={option.placeholder}
        onChange={(param) => eventOnChange(param.target.value)}
        onFocus={() => {
          eventOnChange(search);
          setIsVisible(true); 
        }}
        onBlur={() => {
          setTimeout(() => {
            setIsVisible(false);
          }, 100);  
        }}
      />
      {option.icon ? (
        <div className="absolute top-0 right-0">
          {option.icon}
        </div>
      ) : (<></>)}
      {isVisible && options ? (
        <OptionSearch
          countries={options || []}
          onClick={(param: ICountrySelect) => {
            setSearch(param.Country);
            setOptions(option.searchOptions);
            setIsVisible(false);
            onChange(param.Slug);
          }}  
        />
      ) : (<div></div>)}
    </div>
  );
}