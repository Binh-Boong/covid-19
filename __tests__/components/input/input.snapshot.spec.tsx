import renderer from "react-test-renderer";

import { Input } from "../../../src/components/input/input";

describe("Input", () => {
  it("renders correctly", () => {
      const card = renderer.create(
        <Input
          value="aa"
          onChange={() => {}}
          option={{
            placeholder: "search",
            searchOptions: [
              {
                Country: "Country",
                Slug: "Country",
                ISO2: "Country"
              }
            ]
          }}
        />
      ).toJSON();
    expect(card).toMatchSnapshot();
  });

  it("renders Empty", () => {
    const card = renderer.create(
      <Input
        value="aa"
        onChange={() => {}}
        option={{
          placeholder: "search",
          searchOptions: []
        }}
      />
    ).toJSON();
    expect(card).toMatchSnapshot();
  });

  it("renders Null", () => {
    const card = renderer.create(
      <Input
        value="aa"
        onChange={() => {}}
        option={{
          placeholder: "search",
          icon: <></>,
          className: "cl-test"
        }}
      />
    ).toJSON();
    expect(card).toMatchSnapshot();
  });
});

export {}