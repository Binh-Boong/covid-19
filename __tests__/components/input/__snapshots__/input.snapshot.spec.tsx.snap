// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input renders Empty 1`] = `
<div
  className="relative w-full"
>
  <input
    className="w-full px-3 py-1 text-gray-700 border border-gray-300 border-solid rounded outline-none focus:border-purple-700 focus:shadow-lg"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    placeholder="search"
    value="aa"
  />
  <div />
</div>
`;

exports[`Input renders Null 1`] = `
<div
  className="relative w-full"
>
  <input
    className="cl-test"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    placeholder="search"
    value="aa"
  />
  <div
    className="absolute top-0 right-0"
  />
  <div />
</div>
`;

exports[`Input renders correctly 1`] = `
<div
  className="relative w-full"
>
  <input
    className="w-full px-3 py-1 text-gray-700 border border-gray-300 border-solid rounded outline-none focus:border-purple-700 focus:shadow-lg"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    placeholder="search"
    value="aa"
  />
  <div />
</div>
`;

exports[`Option Search renders Empty 1`] = `
<div
  className="relative w-full"
>
  <input
    className="w-full px-3 py-1 text-gray-700 border border-gray-300 border-solid rounded outline-none focus:border-purple-700 focus:shadow-lg"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    placeholder="search"
    value="aa"
  />
  <div />
</div>
`;

exports[`Option Search renders Null 1`] = `
<div
  className="relative w-full"
>
  <input
    className="cl-test"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    placeholder="search"
    value="aa"
  />
  <div
    className="absolute top-0 right-0"
  />
  <div />
</div>
`;

exports[`Option Search renders correctly 1`] = `
<div
  className="relative w-full"
>
  <input
    className="w-full px-3 py-1 text-gray-700 border border-gray-300 border-solid rounded outline-none focus:border-purple-700 focus:shadow-lg"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    placeholder="search"
    value="aa"
  />
  <div />
</div>
`;
